### PR TITLE
Standardized ETH and ERC20 Gateways

### PR DESCRIPTION
**Pull request by ben-chain**
## Description
This draft PR builds on #219, implementing the gateway contracts and extending them for use with ETH in particular.

Note: We do have plans to add a `withdrawAndCall` function, but probably gonna merge this in without before doing that.

## Questions
- Are the events indexed how we want them to be?  (cc @gigamesh)

### Fixes
- Fixes roadmap567

## Contributing Agreement

- [x] I have read and understood the [Optimism Contributing Guide and Code of Conduct](https://github.com/ethereum-optimism/optimism-monorepo/blob/master/.github/CONTRIBUTING.md) and am following those guidelines in this pull request.

### CODE REVIEW

File: bin/deploy.js [hunk 1]
```diff
@@ -30,6 +30,7 @@ const L2_CROSS_DOMAIN_MESSENGER_ADDRESS =
   env.L2_CROSS_DOMAIN_MESSENGER_ADDRESS || '0x4200000000000000000000000000000000000007';
 let RELAYER_ADDRESS = env.RELAYER_ADDRESS || '0x0000000000000000000000000000000000000000';
 const RELAYER_PRIVATE_KEY = env.RELAYER_PRIVATE_KEY;
+const L1_WETH_ADDRESS = env.L1_WETH_ADDRESS || '0x0000000000000000000000000000000000000000';
```
Comment (K-Ho) [inline bin/deploy.js:33]: Remove

File: bin/deploy.js [hunk 2]
```diff
@@ -91,13 +92,11 @@ const RELAYER_PRIVATE_KEY = env.RELAYER_PRIVATE_KEY;
       maxGasPerQueuePerEpoch: MAX_GAS_PER_QUEUE_PER_EPOCH,
       secondsPerEpoch: SECONDS_PER_EPOCH
     },
+    l1WethAddress: L1_WETH_ADDRESS,
```
Comment (K-Ho) [inline bin/deploy.js:95]: Remove

File: contracts/optimistic-ethereum/OVM/bridge/ETH/OVM_L1ETHGateway.sol [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        onlyFromCrossDomainAccount(l2ERC20Gateway)
+    {
+        _safeTransferETH(_to, _amount);
+
+        emit WithdrawalFinalized(_to, _amount);
+    }
+
+    function _safeTransferETH(address to, uint256 value) internal {
+        (bool success, ) = to.call{value: value}(new bytes(0));
+        require(success, 'TransferHelper::safeTransferETH: ETH transfer failed');
```
Comment (K-Ho) [inline contracts/optimistic-ethereum/OVM/bridge/ETH/OVM_L1ETHGateway.sol:123]:
```suggestion
        require(success, 'safeTransferETH: ETH transfer failed');
```

**Review by K-Ho — COMMENTED** [review]
looking good!

File: src/contract-deployment/config.ts [hunk 4]
```diff
@@ -110,6 +108,13 @@ export const makeContractDeployConfig = async (
         )
       },
     },
+    OVM_L1ETHGateway: {
+      factory: getContractFactory('OVM_L1ETHGateway'),
+      params: [
+        AddressManager.address,
+        '0x4200000000000000000000000000000000000006'
```
Comment (platocrat) [inline src/contract-deployment/config.ts:115]: You need to add a comma ` , ` at the end of line 115  to resolve the CI errors

**Review by platocrat — COMMENTED** [review]
the checks are failing on `yarn lint:typescript` for both Node 10 and Node 11:

Screen Shot 2021-02-16 at 5 25 22 PM

Screen Shot 2021-02-16 at 5 25 40 PM

File: contracts/optimistic-ethereum/iOVM/bridge/ERC20/iOVM_L2ERC20Gateway.sol [hunk 5]
```diff
@@ -0,0 +1,29 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >0.5.0 <0.8.0;
```
Comment (justinjmoses) [inline contracts/optimistic-ethereum/iOVM/bridge/ERC20/iOVM_L2ERC20Gateway.sol:2]: any reason to not allow v0.8?

Comment (platocrat) [reply contracts/optimistic-ethereum/iOVM/bridge/ERC20/iOVM_L2ERC20Gateway.sol:2]: i think their forked version of `solc` only supports up to 0.7.4 atm

Comment (ben-chain) [reply contracts/optimistic-ethereum/iOVM/bridge/ERC20/iOVM_L2ERC20Gateway.sol]: no reason not to remove from the interface though, thanks justin!

File: contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L1ETHGateway.sol [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+        _initiateDeposit(msg.sender, msg.sender);
+    }
+
+    /**
+     * @dev deposit an amount of ERC20 to a recipients's balance on L2
+     * @param _to L2 address to credit the withdrawal to
+     */
+    function depositTo(address _to) external override payable {
+        _initiateDeposit(msg.sender, _to);
+    }
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L1ETHGateway.sol:64]: Is it a standard practice to override here just to add payable? / this doesn't seem to have the standard formatting

File: contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L1ETHGateway.sol [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+    {
+        _safeTransferETH(_to, _amount);
+
+        emit WithdrawalFinalized(_to, _amount);
+    }
+
+    function _safeTransferETH(address to, uint256 value) internal {
+        (bool success, ) = to.call{value: value}(new bytes(0));
+        require(success, 'TransferHelper::safeTransferETH: ETH transfer failed');
+    }
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L1ETHGateway.sol:120]: Nit: non-standard styling on these function params

File: contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L1ETHGateway.sol [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+
+    receive() external payable {
+    }
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L1ETHGateway.sol:123]: Is this `receive(..)` function supposed to be empty?

File: contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+ * This contract mints new tokens when it hears about deposits into the L1 ERC20 gateway.
+ * This contract also burns the tokens intended for withdrawal, informing the L1 gateway to release L1 funds.
+ *
+ * Compiler used: optimistic-solc
+ * Runtime target: OVM
+ */
+contract OVM_L2DepositedERC20 is iOVM_L2DepositedERC20, UniswapV2ERC20, OVM_CrossDomainEnabled {
+    
+    /******************
+     * Contract Events*
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol:27]: Nit: Missing space aver `Events`

File: contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+        uint8 _decimals,
+        string memory _name,
+        string memory _symbol
+    )
+        public
+        OVM_CrossDomainEnabled(_l2CrossDomainMessenger)
+        UniswapV2ERC20(_decimals, _name, _symbol)
+    {}
+
+    // Should we be inheriting this initialization logic?
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol:59]: Where would you inherit from? Are you referring to `OVM_CrossDomainEnabled` ?

Comment (karlfloersch) [reply contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol:59]: Ah or you're talking about using some standard `InitializableContract` thing?

If that's the case IMO don't worry about inheriting it. I like the clarity of having everything in one spot so you're not tracking down various inherited contracts

File: contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol [hunk 11]
```diff
@@ -0,0 +1,159 @@
...
+     * @param _to Address to receive the withdrawal at
+     * @param _amount Amount of the ERC20 to withdraw
+     */
+    function finalizeDeposit(address _to, uint _amount) external override onlyInitialized()
+        onlyFromCrossDomainAccount(address(l1ERC20Gateway))
+    {
+        _mint(_to, _amount);
+        emit DepositFinalized(_to, _amount);
+    }
+}
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol:159]: Nit: There are a bunch of functions in this file with non-standard styling. -- finalizeDeposit, _initiateWithdrawal, withdrawTo

File: contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol [hunk 12]
```diff
@@ -0,0 +1,159 @@
...
+            iOVM_L1ERC20Gateway.finalizeWithdrawal.selector,
+            _to,
+            _amount
+        );
+
+        // Send message up to L1 gateway
+        sendCrossDomainMessage(
+            address(l1ERC20Gateway),
+            data,
+            8999999 // TODO: meter and set with some buffer (actually I think its currently unused in this direction)
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol:135]: Would be good to inherit this value from the EM. And maybe do something like `ExecutionManager.GAS_LIMIT - 1` or something

File: contracts/optimistic-ethereum/iOVM/bridge/tokens/iOVM_L2DepositedERC20.sol [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+     * Public Functions *
+     ********************/
+    function withdraw(uint _amount) external;
+    function withdrawTo(address _to, uint _amount) external;
+    
+    /*************************
+     * Cross-chain Functions *
+     *************************/
+    function finalizeDeposit(address _to, uint _amount) external;
+}
```
Comment (karlfloersch) [inline contracts/optimistic-ethereum/iOVM/bridge/tokens/iOVM_L2DepositedERC20.sol:29]: Note that a lot of these interfaces don't use the standard function naming style, but imo it's probably fine to keep it this way for the interface contracts as long as the function name doesn't get too long

File: src/contract-deployment/config.ts [hunk 14]
```diff
@@ -215,11 +219,8 @@ export const makeContractDeployConfig = async (
     OVM_ETH: {
       factory: getContractFactory('OVM_ETH'),
       params: [
-        AddressManager.address,
-        config.ethConfig.initialAmount,
-        'Ether',
-        18,
-        'ETH',
+        '0x4200000000000000000000000000000000000007',
+        '0x0000000000000000000000000000000000000000', // will be overridden by geth when state dump is ingested.  Storage key: 0x0000000000000000000000000000000000000000000000000000000000000008
```
Comment (karlfloersch) [inline src/contract-deployment/config.ts:223]: We should definitely make a single source of truth for all of these pre-deploy addresses. If we are already hardcoding the addresses then I don't think we need to change it for this PR, but we should definitely at least add an issue for cleaning this up

File: test/contracts/OVM/bridge/assets/OVM_L1ERC20Gateway.spec.ts [hunk 15]
```diff
@@ -0,0 +1,232 @@
...
+  smockit,
+  MockContract,
+  smoddit,
+  ModifiableContract,
+} from '@eth-optimism/smock'
+
+/* Internal Imports */
+import { NON_ZERO_ADDRESS, ZERO_ADDRESS } from '../../../../helpers'
+
+const HARDCODED_GASLIMIT = 8999999
```
Comment (karlfloersch) [inline test/contracts/OVM/bridge/assets/OVM_L1ERC20Gateway.spec.ts:16]: Probably good to resolve this based on the ExecutionManager. Of course ideally we move the env mgr

File: test/contracts/OVM/bridge/assets/OVM_L2DepositedERC20.spec.ts [hunk 16]
```diff
@@ -0,0 +1,213 @@
...
+  smockit,
+  MockContract,
+  smoddit,
+  ModifiableContract,
+} from '@eth-optimism/smock'
+
+/* Internal Imports */
+import { NON_ZERO_ADDRESS, ZERO_ADDRESS } from '../../../../helpers'
+
+const HARDCODED_GASLIMIT = 8999999
```
Comment (karlfloersch) [inline test/contracts/OVM/bridge/assets/OVM_L2DepositedERC20.spec.ts:16]: Again a good spot to query the EM

**Review by karlfloersch — COMMENTED** [review]
This looks dope! Left a few comments/nits but overall imo we're verrrry close to a merge!

File: contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol (hunk 10 above)
Comment (ben-chain) [reply contracts/optimistic-ethereum/OVM/bridge/tokens/OVM_L2DepositedERC20.sol:59]: That is what I was referring to--sounds good.

**Review by karlfloersch — APPROVED** [review]
LGTM!

**Review by karlfloersch — APPROVED** [review]
Double LGTM

**Pull request merged**